interfaces/builtin/vcio.go

```diff
@@ -0,0 +1,60 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+// https://docs.broadcom.com/doc/12358545
+// https://github.com/raspberrypi/linux/tree/rpi-5.4.y/drivers/char/broadcom
+const vcioSummary = `allows access to VideoCore I/O`
+
+// The raspberry pi allows programming its GPU from userspace via the /dev/vcio
+// device. These operations should be considered privileged since the driver
+// assumes trusted input, therefore require manual connection.
+const vcioBaseDeclarationSlots = `
+  vcio:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+const vcioConnectedPlugAppArmor = `
+# Description: Can access to vcio.
+
+# vcio v1 access to ARM VideoCore (BCM_VCIO) for userspace GPU programming
+/dev/vcio rw,
+/sys/devices/virtual/bcm2708_vcio/vcio/** r,
+/run/udev/data/c248:[0-9]* r,
```

Comment:
I think this should use the range 240-254, as on my pi the major number for the vcio device node is 240,

```
$ udevadm info --name /dev/vcio 
P: /devices/virtual/bcm2708_vcio/vcio
N: vcio
L: 0
E: DEVPATH=/devices/virtual/bcm2708_vcio/vcio
E: DEVNAME=/dev/vcio
E: MAJOR=240
E: MINOR=0
E: SUBSYSTEM=bcm2708_vcio
```

From https://www.kernel.org/doc/Documentation/admin-guide/devices.txt, the range 240-254 major number is for local/experimental use, and my reading of the vcio driver's usage of alloc_chrdev_region (see https://github.com/raspberrypi/linux/blob/rpi-5.4.y/drivers/char/broadcom/vcio.c#L136) is that they allocate a device range and that this particular major number could change on different systems, so if apps need to read this bit of the udev database, it could be any major number in that range.

Reply:
@anonymouse64 - nice catch! I read the kernel source and can confirm this is using dynamic assignment (though per https://www.kernel.org/doc/Documentation/admin-guide/devices.txt the range is 234-254 for character devices (240-254 is for local/experimental block devices). Done